feat: add ability to define to redirect to app with different base path

offgriddev opened the pull request:
## Overview

This PR adds the ability to redirect to other app through the `appPageURL` and `authPageURL` config. On the client-side, when an `AuthAction.REDIRECT_TO_LOGIN` or `AuthAction.REDIRECT_TO_APP` are used, the current redirect logic always uses the next router, which only redirects within the current app. However, in micro-UI scenarios, these apps, while under the same domain, can be hosted in different apps with different base routes.

Currently, we can set these two URL config parameters as either strings or functions. By allowing users to assign an object to these configs, users can specify the metadata for the URL. 

## Changes 

This PR adds support for this scenario by allowing the `authPageURL` and `appPageURL` to take object literals with the following schema:

```javascript
{
   "destination": string,
   "basePath": boolean 
}
```

Due to the increased complexity of the shared redirect logic around locating the destination URL in `withAuthUser` and `withAuthUserTokenSSR`, I've extracted some common redirect functions in `src/redirects`. In doing so, I've unified the error messaging in the two components, which were just slightly different. I've also organized the component `withAuthUser` and separated the now more complex routing logic into a separate `useCallback`. The routing is done through `window.location.assign` if the user is being redirected to another app or `router.route` if the user is being redirect to the current app. The SSR component takes the whole object literal when assigned as it represents the SSR `redirect` object.

I've also cleaned up some documentation and added the favor around the new schema for `authPageURL` and `appPageURL`.

### Checks
- [x] Non-breaking API changes
- [x] Tests

### Issues
- closes #187

prescottprue commented:
👍  Super helpful for us since it addresses #187

offgriddev commented:
@kmjennison Hope you are well. I've reworked this PR to focus on resolving _only_ issue #187. In the PR, I've allowed an object to be set or resolved to `authPageURL` or `appPageURL` that allows  for redirecting to another app through a `window.location.assign` call instead of a call to the next router. I also included a focused refactor of some of the redirect code that I saw duplicated across the SSR and CSR code. One side effect to that is the error messaging changed slightly. The errors were slightly different in the two components, but I found the more generic messaged in the CSR component worked the best, so I kept that one. Let me know what you think or if you have any questions! 👋🏼

prescottprue commented on README.md:
```diff
@@ -270,9 +270,17 @@ Option | Description | Default
+`appPageURL` | The redirect destination URL when we should redirect to the app. Can be a string or a function that receives `{ ctx, AuthUser }` and returns a URL as a string or an object (see below for schema). | `config.appPageURL`
+`authPageURL` | The redirect destination URL when we should redirect to the login page. Can be a string or a function that receives `{ ctx, AuthUser }` and returns a URL as a string or an object (see below for schema). | `config.authPageURL`
+
+When either `appPageURL` or `authPageURL` are set as an object literal or an object literal is returned from a function, the object must have the following schema:
```
Was there a request to make this work with the existing param instead of adding a new one? I ask because in the case of passing an object the naming doesn't feel like it makes a ton of sense (since it isn't just the url at that point right?)

offgriddev replied:
This was in accordance with Kevin's original suggestion [here](https://github.com/gladly-team/next-firebase-auth/issues/187#issuecomment-854164651). After our discussion, the approach became more clear and the repo sets is up already with the overloading of the `authPageURL` and `appPageURL` as a function. Currently, it reads a little weird, but it could be different to make it more dry, something like `appPageURL.location`. It fit more within the current patterns here to allow someone to express the metadata of the url in an object than to add another property to the config.

offgriddev replied:
In the process of looking into a slight schema adjustment, I noticed the SSR component has a bug and there's a section of untested code that didn't catch it. I'll have a fix up for this shortly.

offgriddev replied:
I've made a slight update to the schema and added some test coverage to the SSR component. Let me know if you have any more questions or concerns about the approach here.

prescottprue replied:
Great find - no I'm good, you answered all my questions here.

prescottprue commented on src/redirects.js:
```diff
@@ -0,0 +1,50 @@
+
+  if (!destination) {
+    throw new Error(
+      'The "appPageURL" must be set to a non-empty string, an object literal containing "url" and "basePath", or resolve to either'
```
This should be `authPageUrl` right?

offgriddev replied:
Yes. Thanks!

kmjennison commented:
@offgriddev Thanks for this reworking!

One question I have is how Next.js handles the base path in its server-side redirect returned in [`getServerSideProps`](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering). I couldn't find documentation on this. Do you know if it always uses the base path, never uses the base path, or if that's configurable?

We'd want identical redirect URLs regardless of the redirect happening client-side or server-side. I haven't had a chance to look more closely, but it seems like they'd differ because the server-side redirect doesn't have any base path logic. Is that right?

offgriddev commented:
> One question I have is how Next.js handles the base path in its server-side redirect returned in [`getServerSideProps`](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering). I couldn't find documentation on this. Do you know if it always uses the base path, never uses the base path, or if that's configurable?
> 

@kmjennison it actually looks like `basePath` is an undocumented property for `getServerSideProps`: https://github.com/vercel/next.js/blob/59f7676966570093f56ff35fb0d83eb97b394b7b/packages/next/server/base-server.ts#L1899

I’m going to look into this a bit more, but this looks to be the place where the redirect is handled on the server. If this is true, there is a little bit more I can do here in the SSR component.

offgriddev commented:
@kmjennison Let me know what you think of the change I pushed. With the discovery of the `basePath` functionality implemented on the next server, the object literal for `authPageURL` and `appPageURL` could match the redirect object and be passed directly. To me, it cleans things up a bit and reduces the mapping, but let me know your thoughts!

offgriddev commented:
@kmjennison NextJS  responded to my issue quickly, and now the `basePath` is documented for SSR :)

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,50 @@
+
+  if (!destination) {
+    throw new Error(
+      'The "authPageURL" must be set to a non-empty string, an object literal containing "url" and "basePath", or resolve to either'
```
* The "url" should be "destination" instead.
* Just remove "basePath", as it's optional
* Change "resolve to either" to "a function that returns either"

kmjennison replied:
Suggestion: 'The "authPageURL" must be set to a non-empty string, an object containing a "destination" property, or a function that returns either'

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,50 @@
+  if (!authRedirectDestination) {
+    throw new Error(
+      'The "authPageURL" config setting must be set when using `REDIRECT_TO_LOGIN`.'
+    )
```
The two errors here feels redundant. Just rely on the one below.

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,50 @@
+  }
+  const destination = getDestination(ctx, AuthUser, authRedirectDestination)
+
+  if (!destination) {
```
Let's do some simple validation on this to ensure it's a non-empty string or an object with "destination" set to a non-empty string rather than just ensuring truthiness.

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  return destination
+}
```
Please add tests for this module.

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,50 @@
+  return destination
+}
+
+export const getRedirectToAppDestination = (
```
Same changes here as to the function above.

kmjennison commented on src/__tests__/withAuthUserTokenSSR.test.js:
```diff
@@ -553,6 +577,53 @@ describe('withAuthUserTokenSSR: redirect and composed prop logic', () => {
+      },
+    })
+  })
+
```
Please add tests that include `basePath: true`.

kmjennison commented on src/__tests__/withAuthUserTokenSSR.test.js:
```diff
@@ -437,6 +437,30 @@ describe('withAuthUserTokenSSR: redirect and composed prop logic', () => {
+      whenUnauthed: AuthAction.REDIRECT_TO_LOGIN,
+      authPageURL: ({ ctx }) => ({
+        destination: `/my-login?next=${ctx.pathname}`,
+        permanent: false,
```
Remove `permanent` from tests when removing it from the config value.

kmjennison reviewed (changes requested):
Thanks for pushing this forward! It's great that Next supports what we need for this feature.

Requested changes. LMK if you have any questions.

offgriddev commented:
@kmjennison Thanks for the review! I’ll get a revision up later today :)

kmjennison commented:
With the greater flexibility in the authPageURL and appPageURL types, we'll also have to update the config section of the README and the TypeScript definitions.

offgriddev commented on src/withAuthUser.js:
```diff
@@ -118,48 +123,41 @@ const withAuthUser =
+          ) {
+            window.location.assign(redirectURL.destination)
+          } else {
+            router.replace(redirectURL)
```
Good catch! 🎸

offgriddev commented on src/redirects.js:
```diff
@@ -0,0 +1,50 @@
+
+  return redirectDestination
+}
+
```
Great callouts here. Some lended to deeper complexities, but also a cleaner design. I'll have this mostly done and will get the remainder on Monday.

offgriddev commented:
@kmjennison 👋🏼  I hope your Monday is going well. I worked through your feedback. I just pushed some changes for your review when you get a chance. 

In adding tests for the redirect module, I moved the logic for getting the item from the component config or the global config for the redirect. This was the code that looked like this `const redirectDestination = appPageURL || getConfig().appPageURL`. This allowed the caller to only need `getRedirectToLoginDestination({ AuthUser, ctx })`. 

I also updated the TS and the README. Let me know if you want me to add any additional context there. Let me know if you'd like me to make any further adjustments!

kmjennison commented on index.d.ts:
```diff
@@ -48,9 +48,14 @@ interface AuthUserContext extends AuthUser {
+  basePath: boolean
+}
+
+type PageURL = string | RedirectObject | URLResolveFunction
```
Looks great!

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,79 @@
+}
+
+export const getRedirectToLoginDestination = ({
+  options,
```
It looks like you're only using one property of this options object? In that case, just pass that value, like "urlSetting". The "options" object is hard to reason about.

offgriddev replied:
Yeah, I totally see that. Thanks for catching it!

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,79 @@
+
+const getRedirectByUrlConfigName = (redirectConfig) => {
+  const { redirectConfigName, options, ctx, AuthUser } = redirectConfig
+  return throwWhenInvalid(
```
Suggestion for clarity: have the `throwWhenInvalid` return void and simply call it before returning the result of `getDestination`.

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,79 @@
+  AuthUser,
+  ctx,
+} = {}) =>
+  getRedirectByUrlConfigName({
```
Suggestion for clarity: get the default URL from the config here, along with the urlSetting, and pass those—or even just pass the relevant one (config value if urlSetting is not defined). That would be easier to reason about rather than relying on config key names.

offgriddev replied:
The only side effect here is that the error message relies knowing the prop name, but maybe we can move this validation up so where we don't need to reason about a config setting name.

offgriddev replied:
But, I mean we could rethink the error message and remove the complexity.

offgriddev replied:
@kmjennison What are your thoughts on updating this error message for the new context of being within the redirect module? Something like `The "redirectDestination" must be set to a non-empty string ...`? The caller will know what it means because it needs to be explicitly set in the props for the function.

kmjennison replied:
@offgriddev I think it's still fine to pass the "redirectConfigName" value for use only in error messages.

offgriddev replied:
Sounds good. I'll work this in today.

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  return null
+}
```
This looks great.

kmjennison commented on src/redirects.js:
```diff
@@ -0,0 +1,79 @@
+    ctx,
+  })
+
+export const getRedirectToAppDestination = ({ options, AuthUser, ctx } = {}) =>
```
Same comments as above.

kmjennison commented on src/withAuthUser.js:
```diff
@@ -124,48 +127,34 @@ const withAuthUser =
-        }
+        const destination = getRedirectToAppDestination({
+          AuthUser,
+          options: withAuthUserConfig,
```
Please explicitly pass only the properties that are used (construct a new object or name each property).

kmjennison commented on src/withAuthUserTokenSSR.js:
```diff
@@ -93,54 +90,27 @@ const withAuthUserTokenSSR =
+      const redirect = getRedirectToLoginDestination({
+        ctx,
+        AuthUser,
+        options: withAuthUserSSRConfig,
```
Please explicitly pass only the properties that are used (construct a new object or name each property).

kmjennison commented on src/__tests__/redirects.test.js:
```diff
@@ -0,0 +1,208 @@
+      })
+    })
+  })
+})
```
Nice work on these tests.

kmjennison commented on src/__tests__/withAuthUser.test.js:
```diff
@@ -897,6 +900,112 @@ describe('withAuthUser: rendering/redirecting', () => {
+    expect(window.location.assign).toHaveBeenCalledWith('/some-auth-page')
+  })
+
+  it('calls the "appPageURL" function with an undefined context and unauthed AuthUser if redirecting to the login outside the base path on client', () => {
```
Fix test description. This is testing the object URL, not a function.

offgriddev replied:
This one was both 😆

kmjennison commented on src/__tests__/withAuthUser.test.js:
```diff
@@ -897,6 +900,112 @@ describe('withAuthUser: rendering/redirecting', () => {
+    expect(window.location.assign).toHaveBeenCalledWith('/my-app/here/')
+  })
+
+  it('calls the "appPageURL" function with an undefined context and unauthed AuthUser if redirecting to the login within the base path on client', () => {
```
Fix test description. This is testing the object URL, not a function.

kmjennison commented on src/__tests__/withAuthUser.test.js:
```diff
@@ -897,6 +900,112 @@ describe('withAuthUser: rendering/redirecting', () => {
     expect(container.firstChild).toBeNull()
   })
+
+  it('calls the "authPageURL" function with an undefined context and unauthed AuthUser if redirecting to the login outside the base path on client', () => {
```
Also duplicate this test for appPageURL.

kmjennison reviewed (changes requested):
@offgriddev Thanks for these changes. Another round of feedback. A few notes:
* Some feedback (esp for redirect logic) is opinionated about code clarity. If you'd prefer to ignore it, I can follow up with a PR to take care of those changes.
* Please push additional commits without squashing. That helps with code review on GitHub to compare what's changed since the last review.
* Nice work on the tests!

offgriddev commented on README.md:
```diff
@@ -297,8 +300,20 @@ It accepts the following options:
+{
+  destination: string, // the URL destination of a redirect
+  basePath: boolean, // Optional, defaults to true. Whether to use the Next.js base path.
+}
```
Oh thanks! I misunderstood your comment about this.

offgriddev commented on src/redirects.js:
```diff
@@ -0,0 +1,90 @@
+    getDestination({
+      ctx,
+      AuthUser,
+      redirectDestination: redirectURL || getConfig()[redirectConfigName],
```
This logic really cleaned up

offgriddev reviewed (commented):
@kmjennison I just finished working through the feedback. Let me know if you have anything else, but I'm feeling good about these changes. Thanks for the helpful comments and suggestions on this PR!

offgriddev commented:
@kmjennison I hope you're well. Is there any chance we can get this merged into the next release, or do you think there are some changes required still? Let me know. Happy Holidays to you if I don't hear back from you before next week.

kmjennison reviewed (approved):
Looks great! @offgriddev thanks for the good work and iterations. I'll get this out in a release soon.

offgriddev commented:
@kmjennison my pleasure! It was great to work with you through it!
